### ISSUE TITLE
redirect doesn't cancel or consume current response

### ISSUE THREAD

**Issue by ronag** [whatwg/fetch#1288]
Even though a server shouldn't send a body with a redirect response, it is still possible. The way the spec is defined we have a deadlock here if the connection is re-used where the redirect response body is waiting to be consumed but we discard the reference and start a new request (mainFetch).

**Comment by ronag** [comment]
I think you want to cancel/end/abort the response stream in some way regardless of http version? Maybe just write "cancel response stream" and leave what that means as an implementation detail?

**Comment by annevk** [comment]
Step 8 of https://fetch.spec.whatwg.org/#http-fetch does something like that. #637 is about making that version-agnostic.

This request is slightly different I suppose in that you are asking for mandatory behavior which is not unreasonable. We'd need everyone to agree on what that behavior should be though.

**Comment by ronag** [comment]
Ah, you mean:

> If actualResponse’s status is not 303, request’s body is not null, and the connection uses HTTP/2, then user agents may, and are even encouraged to, transmit an RST_STREAM frame.

Not sure why the "status is not 303" pre-condition exists? A 303 response may also contain a body if the server doesn't follow spec.

**Comment by ronag** [comment]
In terms of "needs implementer interest" I can add Node to that list.

**Comment by domenic** [comment]
The WHATWG working mode is about implementing specs in browser engines, so we need multiple browser engines interested in such a thing.

**Comment by domenic** [comment]
Yeah, you can learn more about the WHATWG here: https://whatwg.org/faq#what-is-the-whatwg

**Comment by ronag** [comment]
Likewise the 407 case does not cleanup the active response.
